docs/adr/adr-009-telemetry.md

```diff
@@ -0,0 +1,466 @@
+  * Push based
+  * Widely supported option in the Linux community
+  * Written in C
+  * Decade of experience optimized to bare metal
+  * Perfect for local monitoring setups
+  * Unfortunately, does not support OTLP
+* [Uptrace](https://get.uptrace.dev/guide/#what-is-uptrace)
+  * The most recent (~1 year)
+  * Tight to Clickhouse DB
+  * The most lightweight
+  * Supports OTLP
+  * OSS and can be deployed locally
+  * Provides hosted solution
+* Prometheus+Graphana
+  * Pull based
+  * No native OTLP support
+    * Thought there [spec](https://github.com/open-telemetry/wg-prometheus/blob/main/specification.md) to fix this
+  * Still, can be used with Otel Collector
+
+Similarly, no strictness around backend solution with only OTLP support requirement. Natively or through OTLP exporter.
+
+## Design
+
+### Tracing Design
+
+Tracing allows to see _how_ any process progresses through different modules, APIs and networks, as well as timings of
+each operation and any events or errors as they occur.
+
+A visual example of a generic tracing dashboard provided via [Uptrace](https://uptrace.dev/) backend
+![tracing](img/tracing-dashboard.png)
+
+Mainly, for `ShrEx` and reconstruction analysis we need to know if the reconstruction succeeded and the time it took for
+the big block sizes(EDS >= 128). The tracing in this case would provide all the data for the whole reconstruction
+operation and for each sub operation within reconstruction, e.g time spent specifically on erasure coding
+> NOTE: The exact compute time is not available unless [rsmt2d#107](https://github.com/celestiaorg/rsmt2d/issues/107)
+> is fixed.
+
+#### Spans
+
+Span represents an operation (unit of work) in a trace. They keep the time when operation _started_ and _ended_. Any
+additional user defined _attributes_, operation status(success or error with an error itself) and events/logs that
+may happen during the operation.
+
+Spans also form a parent tree, meaning that each span associated to a process can have multiple sub processes or child
+spans and vice-versa. Altogether, this feature allows to see the whole trace of execution of any part of the system, no
+matter how complex it is. This is exactly what we need to analyze our reconstruction performance.
+
+#### Tracing Integration Example
+
+First, we define global pkg level tracer to create spans from within `ipld` pkg. Basically, it groups spans under
+common logical namespace and extends the full name of each span.
+
+```go
+var tracer = otel.Tracer("ipld")
+```
+
+Then, we define a root span in `ipld.Retriever`:
+
+```go
+import "go.opentelemetry.io/otel"
+
+func (r *Retriever) Retrieve(ctx context.Context, dah *da.DataAvailabilityHeader) (*rsmt2d.ExtendedDataSquare, error) {
+  ctx, span := tracer.Start(ctx, "retrieve-square")
+  defer span.End()
+
+  span.SetAttributes(
+    attribute.Int("size", len(dah.RowsRoots)),
+    attribute.String("data_hash", hex.EncodeToString(dah.Hash())),
+  )
```

Comment:
Cardinality issue right here using datahash as attribute. The topic of high cardinality worth its own paragraph as it seems it happens to be encountered by devs from time to time.